src/linear_algebra/quotient.lean

```diff
@@ -100,18 +101,48 @@ instance : add_comm_group (quotient p) :=
     refl },
   gsmul_neg' := by { rintros n ⟨x⟩, simp_rw [gsmul_neg_succ_of_nat, gsmul_coe_nat], refl }, }
 
-instance : has_scalar R (quotient p) :=
+section has_scalar
+
+variables {S : Type*} [has_scalar S R] [has_scalar S M] [is_scalar_tower S R M] (P : submodule R M)
+
+instance : has_scalar S (quotient P) :=
```

Comment:
This is actually a dangerous instance together with [`has_scalar.comp.is_scalar_tower`](https://leanprover-community.github.io/mathlib_docs/find/has_scalar.comp.is_scalar_tower/): the `is_scalar_tower` parameter gets inferred while only the `has_scalar R M` parameter is not a metavar, which means the `has_scalar.comp` subexpression of `comp.is_scalar_tower` always unifies, and you get an infinite loop. See also the output of the `fails_quickly` linter.

The correct solution is to require that `has_scalar.comp.is_scalar_tower` can't be applied before the `has_scalar` instance has been found, but that probably requires some tweaks in the elaborator. Otherwise, either this instance or `has_scalar.comp.is_scalar_tower` has to go. And I'm inclined to get rid of the latter, because I'm sure we'll see this issue come up later again.

Reply:
Hmm, that's not the model I had in my head of how `is_scalar_tower` worked. I assumed that it always found all the `has_scalar` instance first, and then typeclass search was just responsible for the task "find a proof that these instances are compatible".